File: cpp/tests/test_ima.cpp
/// IMA tests: ES hand cases, sqrt(10) scaling, LH-ladder monotonicity and
/// collapse property, category-sum validation, backtest zone/multiplier table
/// edges, SES and IMA capital error paths.

#include <gtest/gtest.h>

#include <cmath>

#include "frtb/ima.hpp"
#include "test_helpers.hpp"

namespace {

using namespace frtb;

TEST(ExpectedShortfall, HandCase) {
    // n=4, alpha=0.975 -> k = ceil(0.1) = 1: the single worst loss.
    std::vector<double> pnl = {-10.0, 5.0, -30.0, 2.0};
    EXPECT_DOUBLE_EQ(expected_shortfall_daily(pnl), 30.0);
    // n=40, alpha=0.975 -> k = 1 (0.025*40 = 1.0 with the epsilon guard).
    std::vector<double> p40(40, 1.0);
    p40[7] = -99.0;
    EXPECT_DOUBLE_EQ(expected_shortfall_daily(p40), 99.0);
    // n=260 -> k = 7 (asserted indirectly by the golden ES cases).
    EXPECT_DOUBLE_EQ(es_base_10d(pnl), std::sqrt(10.0) * 30.0);
}

TEST(ExpectedShortfall, Errors) {
    EXPECT_THROW(expected_shortfall_daily({}), std::invalid_argument);
    EXPECT_THROW(expected_shortfall_daily({1.0}, 1.5), std::invalid_argument);
    EXPECT_THROW(expected_shortfall_daily({1.0, NAN}), std::invalid_argument);
}

TEST(LhLadder, SingleCategoryCollapse) {
    // One category with horizon LH collapses to ES1 * sqrt(LH/10).
    const SbmParams& params = frtb_test::params();
    std::vector<double> pnl;
    for (int i = 0; i < 40; ++i) pnl.push_back((i % 7) - 3.0 - (i == 5 ? 50.0 : 0.0));
    const double es1 = es_base_10d(pnl, params.ima_alpha);
    for (const auto& [cat, lh] : params.category_lh) {
        CategoryPnl cats = {{cat, pnl}};
        double es_lh =
            es_lh_scaled(pnl, cats, params.category_lh, params.lh_ladder, params.ima_alpha);
        EXPECT_NEAR(es_lh, es1 * std::sqrt(lh / 10.0), 1e-10) << "category " << cat;
        EXPECT_GE(es_lh, es1 - 1e-12);  // ladder is monotone
    }
}

TEST(LhLadder, MonotoneOnBundledDesks) {
    const frtb::Results& res = frtb_test::results();
    for (const auto& [d, r] : res.ima) {
        EXPECT_GE(r.es_lh, r.es_base) << d;
        EXPECT_GE(r.imcc, 0.0) << d;
    }
}

TEST(LhLadder, ValidationErrors) {
    const SbmParams& params = frtb_test::params();
    std::vector<double> pnl = {1.0, -2.0, 3.0, -4.0};
    // Category series must sum to the full P&L.
    CategoryPnl bad_sum = {{"ir", {1.0, -2.0, 3.0, -3.0}}};
    EXPECT_THROW(es_lh_scaled(pnl, bad_sum, params.category_lh, params.lh_ladder),
                 std::invalid_argument);
    // Unknown category horizon.
    CategoryPnl unknown = {{"weather", pnl}};
    EXPECT_THROW(es_lh_scaled(pnl, unknown, params.category_lh, params.lh_ladder),
                 std::invalid_argument);
    // Length mismatch.
    CategoryPnl short_series = {{"ir", {1.0, -2.0}}};
    EXPECT_THROW(es_lh_scaled(pnl, short_series, params.category_lh, params.lh_ladder),
                 std::invalid_argument);
    // Ladder must start at 10 and be strictly increasing.
    CategoryPnl ok = {{"ir", pnl}};
    EXPECT_THROW(es_lh_scaled(pnl, ok, params.category_lh, {20, 40}), std::invalid_argument);
    EXPECT_THROW(es_lh_scaled(pnl, ok, params.category_lh, {10, 40, 20}), std::invalid_argument);
}

TEST(Backtest, ZoneAndMultiplierTableEdges) {
    const SbmParams& params = frtb_test::params();
    // Table edges: 4 -> green 1.5; 5 -> amber 1.70; 9 -> amber 1.92; 10 -> red 2.0.
    EXPECT_EQ(backtest_zone(0), "green");
    EXPECT_EQ(backtest_zone(4), "green");
    EXPECT_EQ(backtest_zone(5), "amber");
    EXPECT_EQ(backtest_zone(9), "amber");
    EXPECT_EQ(backtest_zone(10), "red");
    EXPECT_EQ(backtest_zone(13), "red");  // > 12 stays red (cap)
    EXPECT_DOUBLE_EQ(backtest_multiplier(4, params), 1.5);
    EXPECT_DOUBLE_EQ(backtest_multiplier(5, params), 1.70);
    EXPECT_DOUBLE_EQ(backtest_multiplier(6, params), 1.75);
    EXPECT_DOUBLE_EQ(backtest_multiplier(9, params), 1.92);
    EXPECT_DOUBLE_EQ(backtest_multiplier(10, params), 2.0);
    EXPECT_DOUBLE_EQ(backtest_multiplier(50, params), 2.0);
    EXPECT_THROW(backtest_zone(-1), std::invalid_argument);
}

TEST(Backtest, StrictExceptionCounting) {
    const SbmParams& params = frtb_test::params();
    // PnL == -VaR is NOT an exception (strict inequality).
    BacktestResult r = backtest({-10.0, -10.1, 5.0}, {10.0, 10.0, 10.0}, params);
    EXPECT_EQ(r.exceptions, 1);
    EXPECT_EQ(r.zone, "green");
    EXPECT_DOUBLE_EQ(r.multiplier, 1.5);
    EXPECT_THROW(backtest({1.0}, {1.0, 2.0}, params), std::invalid_argument);
    EXPECT_THROW(backtest({}, {}, params), std::invalid_argument);
    EXPECT_THROW(backtest({1.0}, {-1.0}, params), std::invalid_argument);
}

TEST(Ses, SumAndErrors) {
    EXPECT_DOUBLE_EQ(ses({{"f1", "d1", 100.0}, {"f2", "d2", 50.5}}), 150.5);
    EXPECT_DOUBLE_EQ(ses({}), 0.0);
    EXPECT_THROW(ses({{"f1", "d1", -1.0}}), std::invalid_argument);
}

TEST(ImaCapital, FormulaAndErrors) {
    EXPECT_DOUBLE_EQ(ima_capital(100.0, 1.5, 20.0, 5.0), 175.0);
    EXPECT_THROW(ima_capital(-1.0, 1.5, 0.0), std::invalid_argument);
    EXPECT_THROW(ima_capital(1.0, 1.5, -1.0), std::invalid_argument);
}

}  // namespace

File: python/tests/test_ima.py
"""IMA sketch: ES definition, LH ladder, IMCC, backtest zones/multipliers, SES."""
import math

import pytest

from frtb.ima import (backtest, backtest_multiplier, backtest_zone, es_base_10d,
                      es_lh_scaled, expected_shortfall_daily, ima_capital, imcc,
                      ses)


class TestExpectedShortfall:
    def test_pinned_tail_definition(self):
        # n = 40, alpha = 0.975 -> k = ceil(1) = 1 -> ES = worst loss
        pnl = [float(i) for i in range(-20, 20)]
        assert expected_shortfall_daily(pnl, 0.975) == 20.0
        assert es_base_10d(pnl, 0.975) == pytest.approx(20.0 * math.sqrt(10.0), abs=1e-12)

    def test_k_of_two(self):
        # n = 80 -> k = ceil(2) = 2 -> mean of two worst losses
        pnl = [0.0] * 78 + [-30.0, -10.0]
        assert expected_shortfall_daily(pnl, 0.975) == pytest.approx(20.0, abs=1e-12)

    def test_errors(self):
        with pytest.raises(ValueError, match="empty"):
            expected_shortfall_daily([], 0.975)
        with pytest.raises(ValueError, match="alpha"):
            expected_shortfall_daily([1.0], 1.5)
        with pytest.raises(ValueError, match="finite"):
            expected_shortfall_daily([1.0, float("inf")], 0.975)


class TestLhLadder:
    LADDER = (10, 20, 40, 60, 120)

    def test_single_category_sqrt_scaling(self):
        # one category with LH = 40: ladder collapses to sqrt(40/10) = 2x base
        pnl = [math.sin(3.0 * i) * 100.0 for i in range(100)]
        es_b = es_base_10d(pnl)
        es_l = es_lh_scaled(pnl, {"fx": pnl}, {"fx": 40}, self.LADDER)
        assert es_l == pytest.approx(2.0 * es_b, abs=1e-9)

    def test_lh10_category_equals_base(self):
        pnl = [math.cos(2.0 * i) * 50.0 for i in range(80)]
        es_l = es_lh_scaled(pnl, {"x": pnl}, {"x": 10}, self.LADDER)
        assert es_l == pytest.approx(es_base_10d(pnl), abs=1e-9)

    def test_ladder_monotone(self):
        # property-style loop: LH scaling never decreases the ES, and it is
        # non-decreasing in the category liquidity horizon
        pnl = [math.sin(1.7 * i) * 80.0 + math.cos(0.3 * i) * 40.0 for i in range(120)]
        base = es_base_10d(pnl)
        prev = 0.0
        for lh in (10, 20, 40, 60, 120):
            es_l = es_lh_scaled(pnl, {"c": pnl}, {"c": lh}, self.LADDER)
            assert es_l >= base - 1e-12
            assert es_l >= prev - 1e-12
            prev = es_l

    def test_two_categories_hand(self):
        # cat a (LH 10) + cat b (LH 20):
        # ES_LH^2 = ES(full)^2 + ES(b)^2 * (20-10)/10
        a = [math.sin(2.1 * i) * 30.0 for i in range(100)]
        b = [math.cos(1.3 * i) * 60.0 for i in range(100)]
        full = [x + y for x, y in zip(a, b)]
        got = es_lh_scaled(full, {"a": a, "b": b}, {"a": 10, "b": 20}, self.LADDER)
        want = math.sqrt(es_base_10d(full) ** 2 + es_base_10d(b) ** 2)
        assert got == pytest.approx(want, abs=1e-9)

    def test_category_sum_mismatch_raises(self):
        with pytest.raises(ValueError, match="sum"):
            es_lh_scaled([1.0, -2.0, 3.0], {"a": [1.0, 1.0, 1.0]}, {"a": 20},
                         self.LADDER)

    def test_missing_category_lh_raises(self):
        with pytest.raises(ValueError, match="liquidity horizon"):
            es_lh_scaled([1.0, -1.0], {"a": [1.0, -1.0]}, {}, self.LADDER)

    def test_bad_ladder_raises(self):
        with pytest.raises(ValueError, match="10d base"):
            es_lh_scaled([1.0, -1.0], {"a": [1.0, -1.0]}, {"a": 20}, (20, 40))


class TestImcc:
    def test_imcc_rho_blend(self, params):
        # single category: full ES == partial ES -> IMCC == ES_LH exactly
        pnl = [math.sin(0.9 * i) * 100.0 for i in range(90)]
        want = es_lh_scaled(pnl, {"ir": pnl}, params.category_lh, params.lh_ladder,
                            params.ima_alpha)
        assert imcc(pnl, {"ir": pnl}, params) == pytest.approx(want, abs=1e-9)

    def test_imcc_between_full_and_sum(self, params):
        a = [math.sin(2.1 * i) * 30.0 for i in range(100)]
        b = [math.cos(1.3 * i) * 60.0 for i in range(100)]
        full = [x + y for x, y in zip(a, b)]
        cats = {"eq": a, "fx": b}
        v = imcc(full, cats, params)
        es_full = es_lh_scaled(full, cats, params.category_lh, params.lh_ladder)
        es_sum = sum(es_lh_scaled(s, {c: s}, params.category_lh, params.lh_ladder)
                     for c, s in cats.items())
        assert es_full <= v + 1e-9 <= es_sum + 1e-9
        assert v == pytest.approx(0.5 * es_full + 0.5 * es_sum, abs=1e-9)


class TestBacktest:
    def test_zone_edges(self):
        assert backtest_zone(0) == "green"
        assert backtest_zone(4) == "green"
        assert backtest_zone(5) == "amber"
        assert backtest_zone(9) == "amber"
        assert backtest_zone(10) == "red"
        assert backtest_zone(13) == "red"

    def test_multiplier_table_edges(self, params):
        assert backtest_multiplier(4, params) == 1.5
        assert backtest_multiplier(5, params) == 1.70
        assert backtest_multiplier(9, params) == 1.92
        assert backtest_multiplier(10, params) == 2.0
        assert backtest_multiplier(13, params) == 2.0  # > 12: red-zone cap

    def test_exception_counting(self, params):
        pnl = [-5.0, -15.0, 3.0, -10.0, -10.1]
        var = [10.0, 10.0, 10.0, 10.0, 10.0]
        res = backtest(pnl, var, params)
        assert res.exceptions == 2  # strictly below -VaR only (-15, -10.1)
        assert res.zone == "green"
        assert res.multiplier == 1.5

    def test_errors(self, params):
        with pytest.raises(ValueError, match="mismatch"):
            backtest([1.0], [1.0, 2.0], params)
        with pytest.raises(ValueError, match="non-negative"):
            backtest([1.0], [-1.0], params)
        with pytest.raises(ValueError, match="negative"):
            backtest_zone(-1)


class TestSesAndCapital:
    def test_ses_zero_diversification_sum(self):
        entries = [{"factor": "a", "desk": "d", "stressed_loss": 100.0},
                   {"factor": "b", "desk": "d", "stressed_loss": 250.0}]
        assert ses(entries) == 350.0
        assert ses([]) == 0.0

    def test_ses_negative_raises(self):
        with pytest.raises(ValueError, match="stressed_loss"):
            ses([{"factor": "a", "desk": "d", "stressed_loss": -1.0}])

    def test_ima_capital_formula(self):
        assert ima_capital(100.0, 1.7, 30.0, 5.0) == pytest.approx(205.0, abs=1e-12)
        with pytest.raises(ValueError, match="multiplier"):
            ima_capital(100.0, -1.0, 0.0)
